confluent_kafka/serialization/serializers.py

```diff
@@ -0,0 +1,455 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import struct
+
+from .error import SerializerError
+
+
+class SerializerBase(object):
```

Comment:
i don't see any benefit to having this class. the only things that come to mind are:
1. allows the clients to easily check whether the passed in thing is serializer like, but you can do that without a base class. 
2. a good place to document serialize and deserialize, but you could do that in the client class.

overall seems un-pythonic and not necessary.

[disclaimer: all my comments are from the point of view of someone who doesn't really know python, so come with some uncertainty]

Reply:
I disagree I think a base class is the clearest way to express the contract thats expected from an implementation.

I cant say for sure if it is or isn't "pythonic" but its a fairly common pattern used by some of the more popular libraries. I.e. Requests.

Python has actually added similar functionality with its Abstract Base Class implementation but there are some conpatibility issues to contend with. This is a super simple usecase so I opted to avoid dealing with it as it offers very little additional value.

https://docs.python.org/3/library/abc.html

Reply:
can you point to the actual example from `Request`s? my real issue is it didn't seem pythonic (from what I know of python), and think that should be prioritized here over anything else. if there is indeed strong precedent, i have nothing against it.

note: i also don't like the idea of having both serialize and deserialize on the same interface (can't recall if i mentioned that before), as they're never both used together in the same context.

Reply:
Do you care to clarify what about it exactly isn't Pythonic? This very much so feels like the right tool for the job to me given the constraints on the language. The Serializing Producer expects serializers to behave in a certain way, this class clearly documents what that behavior is. 

Requests AuthBase
https://github.com/psf/requests/blob/d79024f246d1037c4196da6027c2c2be08719c8d/requests/auth.py#L72


Requests BaseAdaptor
https://github.com/psf/requests/blob/fd13816d015c4c90ee65297fa996caea6a094ed1/requests/adapters.py#L55

I'm not sure there is much value gained from separating the serializer and deserializer interfaces. The both serve a similar purpose using shared logic and would both need to change under the same conditions.

Reply:
my impression was python people were big into duck typing and not really interfaces, but if they're not, that's fine, my intuition on what's idiomatic is probably off.

on splitting: it's only marginally more effort on your part, and no more effort to use to have both serializers and deserializers. the benefit is an API that's more precise: 1. the user doesn't need to wonder why they're passing a thing called a serializer into the consumer when surely it should be deserializing 2. they don't need to make assumptions on whether or not the producer might use the deserialize method - they know for sure because it's not passed in.

the benefit is arguably marginal - it's unlikely to cause much confusion. but there's no downside so why not do the thing that's structurally better? all the little things add up.

Reply:
you could address point 1 by calling them serdes `SerdeBase` etc, but not point 2.

Reply:
SerdeBase works for me but I think I'll continue combining the two. This also simplifies the handling of delivery reports since the deserializer is already available to my producer.

Reply:
I prefer to keep them together, unless someone develops strong opinions to the contrary I'm leaving as is.

Reply:
If you do that we can't break them up later.
I suggest you do a base Serializer and base Deserializer classes and have both of them inherit from a common base class.
Users will inherit from Serializer or Deserializer, but not the base class (which we can mark as non-public), allow us to split them up later as needed.